File: java/0227-basic-calculator-ii.java
class Solution {
    public int calculate(String s) {
        if(s == null || s.length() == 0){
            return 0;
        }
        char last = '+';
        int calc = 0, tail = 0, curr = 0;
        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if(Character.isDigit(c)){
                curr = curr * 10 + (c - '0');
            }
            if((!Character.isDigit(c) && c != ' ') || i == s.length() - 1){
                if(last == '+'){
                    calc = calc + curr;
                    tail = curr;
                }
                else if(last == '-'){
                    calc = calc - curr;
                    tail = -curr;
                }
                else if(last == '*'){
                    calc = (calc - tail) + (tail * curr);
                    tail = tail * curr;
                }
                else{
                    calc = (calc - tail) + (tail / curr);
                    tail = tail / curr;
                }
                curr = 0;
                last = c;
            }
        }
        return calc;
    }
}

File: cpp/0227-basic-calculator-ii.cpp
class Solution {
public:
    int calculate(string s) {
        long result = 0, last = 0, curr = 0;
        char op = '+';                  
        int n = s.size();
        for (int i = 0; i < n; ++i) {
            char c = s[i];
            if (isdigit(c)) {
                curr = curr * 10 + (c - '0');
            }
            if ((!isdigit(c) && !isspace(c)) || i == n - 1) {
                switch (op) {
                    case '+':
                        result += last;   
                        last = curr;      
                        break;
                    case '-':
                        result += last;
                        last = -curr;
                        break;
                    case '*':
                        last = last * curr;
                        break;
                    case '/':
                        last = last / curr; 
                        break;
                }
                op = c;    
                curr = 0;  
            }
        }
        result += last;  
        return (int)result;
    }
};